Kernel/Graphics/VirtIOGPU/VirtIOFrameBufferDevice.cpp

```diff
@@ -0,0 +1,123 @@
+    , m_gpu(virtio_gpu)
+{
+    auto write_sink_page = MM.allocate_user_physical_page(MemoryManager::ShouldZeroFill::No).release_nonnull();
+    auto num_needed_pages = m_gpu->framebuffer_vm_object().page_count();
+    NonnullRefPtrVector<PhysicalPage> pages;
+    for (auto i = 0u; i < num_needed_pages; ++i) {
+        pages.append(write_sink_page);
+    }
+    m_framebuffer_sink_vmobject = adopt_ref(*new AnonymousVMObject(move(pages)));
+}
+
+VirtIOFrameBufferDevice::~VirtIOFrameBufferDevice()
+{
+}
+
+int VirtIOFrameBufferDevice::ioctl(FileDescription&, unsigned request, FlatPtr arg)
+{
+    REQUIRE_PROMISE(video);
+    switch (request) {
+    case FB_IOCTL_GET_SIZE_IN_BYTES: {
+        auto* out = (size_t*)arg;
+        size_t value = m_gpu->framebuffer_size_in_bytes();
+        if (!copy_to_user(out, &value))
+            return -EFAULT;
+        return 0;
+    }
+    case FB_IOCTL_GET_RESOLUTION:
+    case FB_IOCTL_SET_RESOLUTION: {
+        auto* user_resolution = (FBResolution*)arg;
+        FBResolution resolution;
+        resolution.pitch = m_gpu->framebuffer_pitch();
+        resolution.width = m_gpu->framebuffer_width();
+        resolution.height = m_gpu->framebuffer_height();
+        if (!copy_to_user(user_resolution, &resolution))
+            return -EFAULT;
+        return request == FB_IOCTL_GET_BUFFER ? 0 : -ENOTSUP;
+    }
+    case FB_IOCTL_FLUSH_BUFFER: {
+        FBRect user_dirty_rect;
+        if (!copy_from_user(&user_dirty_rect, (FBRect*)arg))
+            return -EFAULT;
+        VirtIOGPURect dirty_rect {
+            .x = user_dirty_rect.x,
+            .y = user_dirty_rect.y,
+            .width = user_dirty_rect.width,
+            .height = user_dirty_rect.height
+        };
+        m_gpu->flush_dirty_window(dirty_rect);
+        return 0;
+    }
+    default:
+        return -EINVAL;
+    };
+}
+
+KResultOr<Region*> VirtIOFrameBufferDevice::mmap(Process& process, FileDescription&, const Range& range, u64 offset, int prot, bool shared)
+{
+    REQUIRE_PROMISE(video);
+    if (!shared)
+        return ENODEV;
+    if (offset != 0)
+        return ENXIO;
+    if (range.size() != page_round_up(m_gpu->framebuffer_size_in_bytes()))
+        return EOVERFLOW;
+    // We only allow one process to map the region
+    if (m_userspace_mmap_region)
+        return ENOMEM;
+
+    auto vmobject = m_are_writes_active ? m_gpu->framebuffer_vm_object().clone() : m_framebuffer_sink_vmobject;
+
+    auto result = process.space().allocate_region_with_vmobject(
+        range,
+        vmobject.release_nonnull(),
+        0,
+        "VirtIOGPU Framebuffer",
+        prot,
+        shared);
+    if (result.is_error())
+        return result;
+    m_userspace_mmap_region = result.value();
+    return result;
+}
+
+void VirtIOFrameBufferDevice::deactivate_writes()
+{
+    m_are_writes_active = false;
+    if (m_userspace_mmap_region) {
+        m_userspace_mmap_region->set_vmobject(m_framebuffer_sink_vmobject.release_nonnull());
```

Comment:
can we just do this?
```suggestion
        m_userspace_mmap_region->set_writable(false);
```

Reply:
I don't think so? If we make it unwritable, trying to write to it would cause the window server to segfault, right? I think this is the best way to do this, the only issue is that if the windowserver tries to take a screenshot when we are at a tty, it will read in garbage data.

There are some downsides to this approach however, it would be simpler if we created an extra framebuffer that is only used with the tty but that felt a bit like a waste of memory. In the context of multiple monitors (from tomuta's PR), that approach would be still fine as long as we mirror that single framebuffer across all displays, we wouldn't have to worry about maintaining O(n) extra framebuffers..

Reply:
i think the best way to fix this is to stop window server from trying to write to the framebuffer if its not being displayed at all, but i guess thats out of scope for this PR.

Reply:
I think as an extension beyond the scope of this PR, we should make the kernel send certain signals to whichever process has mmapped the framebuffers whenever a switch to/from the graphical TTY is done, with the WindowServer being responsible for not attempting to write when it isn't allowed to (and then also implement the write protection change that you've suggested). Until then however, it would probably be better to do this sinkhole approach.